File: USACO/2020/US_Open/Gold/haircut.cpp
#include <bits/stdc++.h>
using namespace std;
const int N = 1e5 + 10;
using ll = long long;

struct fenwick {
    int st[N];
    int lowbit(int x) {return x & (-x);}
    void update(int a, int b) {
        for(int i = a;i<N;i+=lowbit(i)) st[i] += b;
    }
    int query(int a) {
        int ans = 0;
        for(int i = a;i;i-=lowbit(i)) ans += st[i];
        return ans;
    }
    int query(int l, int r) {
        return query(r) - query(l - 1);
    }
} st;

int n, a[N], b[N];

int main() {
    // freopen("haircut.in", "r", stdin);
    // freopen("haircut.out", "w", stdout);
    ios::sync_with_stdio(0);
    cin.tie(0);
    cin>>n;
    for(int i = 1;i<=n;i++) {
        int x; cin>>x;
        ++x;
        a[x] += st.query(x + 1, N - 1);
        st.update(x, 1);
    }
    ll ans = 0;
    for(int i = 1;i<=n;i++) {
        cout<<ans<<'\n';
        ans += a[i];
    }
    return 0;
}

File: USACO/2020/US_Open/Gold/haircut.java
import java.util.*;
import java.io.*;

public class haircut {
    static BufferedReader br;
    static PrintWriter out;
    static StringTokenizer st;

    static String read() throws IOException {
        while(st == null || !st.hasMoreTokens()) {
            st = new StringTokenizer(br.readLine());
        }
        return st.nextToken();
    }
    static int readInt() throws IOException {
        return Integer.parseInt(read());
    }
    static int n;
    static int[] a;
    static long[] b;
    static fenwick fen;
    public static void main(String[] args) throws IOException{
        br = new BufferedReader(new FileReader("haircut.in"));
        out = new PrintWriter(new BufferedWriter(new FileWriter("haircut.out")));
        n = readInt();
        a = new int[n + 10];
        b = new long[n + 10];
        for(int i = 1;i<=n;++i) a[i] = readInt();
        br.close();
        fen = new fenwick(n + 10);
        for(int i = 1;i<=n;++i) {
            b[a[i] + 1] += fen.query(a[i] + 2, n + 5);
            fen.update(a[i] + 1, 1);
        }
        for(int i = 1;i<=n;++i) {
            b[i] += b[i - 1];
            out.println(b[i - 1]);
        }
        out.close();
    }
}

class fenwick {
    int[] c;
    int n;
    public fenwick(int N) {
        n = N;
        c = new int[n + 10];
    }
    public int query(int a, int b) {
        return query(b) - query(a - 1);
    }
    private int query(int a) {
        int ans = 0;
        for(int i = a;i > 0;i-=i & (-i)) ans += c[i];
        return ans;
    }
    public void update(int a, int b) {
        for(int i = a;i<=n;i+=i & (-i)) c[i] += b;
    }
}
